### PR TITLE
Objects are now obtained from low level state when querying a lifted decision list representing a goal-conditioned policy

### PR DESCRIPTION
**Pull request by wester-is-not-real**
Modified query_ldl so it takes as arguments the objects in the current low-level state.

### CODE REVIEW

File: src/approaches/pg3_approach.py [hunk 1]
```diff
@@ -61,14 +62,18 @@ def _solve(self, task: Task, timeout: int) -> Callable[[State], Action]:
         policy."""
         skeleton = []
         atoms_sequence = []
+        current_objects = set()
         atoms = utils.abstract(task.init, self._initial_predicates)
         atoms_sequence.append(atoms)
+        for key in task.init.data:
```
Comment (tomsilver) [inline src/approaches/pg3_approach.py:68]: you can save a few lines and do:
`current_objects = set(task.init)`
to understand why this works, check out the `__iter__()` method of `State` in structs.py

File: src/approaches/pg3_approach.py [hunk 2]
```diff
@@ -475,7 +480,8 @@ def get_next_state(atoms: _S, ground_nsrt: _A) -> _S:
         )
 
         def policy(atoms: _S) -> Optional[_A]:
-            return utils.query_ldl(ldl, set(atoms), goal)
+            objects = {o for a in atoms | goal for o in a.objects}
```
Comment (tomsilver) [inline src/approaches/pg3_approach.py:483]: this looks like the same dangerous pattern that we want to fix right? can we use the already-defined `objects` variable instead? (so just remove this line). hopefully this doesn't diminish the results you were seeing in painting with box-only, let's check after we merge this

File: src/utils.py [hunk 3]
```diff
@@ -1126,6 +1126,9 @@ def _get_entity_combinations(
         for ent in sorted_entities:
             if ent.is_instance(vt):
                 this_choices.append(ent)
+        # if len(this_choices) == 0:
```
Comment (tomsilver) [inline src/utils.py:1129]: remove the commented out code

File: src/utils.py [hunk 4]
```diff
@@ -1134,7 +1137,8 @@ def _get_entity_combinations(
 def get_object_combinations(objects: Collection[Object],
                             types: Sequence[Type]) -> Iterator[List[Object]]:
     """Get all combinations of objects satisfying the given types sequence."""
-    return _get_entity_combinations(objects, types)
+    penis = _get_entity_combinations(objects, types)
```
Comment (tomsilver) [inline src/utils.py:1140]: 😛

File: src/utils.py [hunk 5]
```diff
@@ -1326,9 +1330,8 @@ def _run_heuristic_search(
         lazy_expansion: bool = False) -> Tuple[List[_S], List[_A]]:
     """A generic heuristic search implementation.
 
-    Depending on get_priority, can implement A*, GBFS, or UCS.
-
-    If no goal is found, returns the state with the best priority.
+    Depending on get_priority, can implement A*, GBFS, or UCS. If no
```
Comment (tomsilver) [inline src/utils.py:1333]: revert this change (just because we don't need it)

File: src/utils.py [hunk 6]
```diff
@@ -1347,6 +1350,8 @@ def _run_heuristic_search(
 
     while len(queue) > 0 and time.time() - start_time < timeout and \
           num_expansions < max_expansions and num_evals < max_evals:
+        # if (method == "gbfs" and best_node_priority==36.0):
```
Comment (tomsilver) [inline src/utils.py:1353]: remove

File: tests/test_structs.py [hunk 7]
```diff
@@ -1041,17 +1041,20 @@ def test_lifted_decision_lists():
 
     atoms = {on_table([cup1]), hand_empty([robot])}
     goal = {on([cup1, plate1])}
+    objects = {o for a in atoms | goal for o in a.objects}
```
Comment (tomsilver) [inline tests/test_structs.py:1044]: so as to not suggest a bad pattern, let's just explicitly create the object set: `objects = {cup1, plate1, ...}`

**Review by tomsilver — COMMENTED** [review]
excellent job finding and fixing this issue! please update the title and description of this PR too, so other readers have context

File: src/approaches/pg3_approach.py (hunk 1 above)
Comment (wester-is-not-real) [reply src/approaches/pg3_approach.py:68]: Oh thats pretty cool. Thanks!

File: tests/test_structs.py (hunk 7 above)
Comment (wester-is-not-real) [reply tests/test_structs.py:1044]: Good idea!

File: src/approaches/pg3_approach.py [hunk 8]
```diff
@@ -61,14 +62,17 @@ def _solve(self, task: Task, timeout: int) -> Callable[[State], Action]:
         policy."""
         skeleton = []
         atoms_sequence = []
+        current_objects = set()
```
Comment (tomsilver) [inline src/approaches/pg3_approach.py:65]: you can remove this now

**Review by tomsilver — APPROVED** [review]
one small thing, otherwise LGTM!
